### ISSUE TITLE
Compile error 'Missing class properties transform' when using Next >=10.0.2

### ISSUE THREAD

**Issue by marklawlor** [expo/expo-cli#3226]
**Summary**

Expo projects using the @expo/next-adapter fail to compile when using the Next server on versions >=10.0.2

**Environment**

```
Expo CLI 4.1.6 environment info:
    System:
      OS: Linux 5.4 Ubuntu 20.04.2 LTS (Focal Fossa)
      Shell: 5.8 - /usr/bin/zsh
    Binaries:
      Node: 12.19.0 - ~/.nvm/versions/node/v12.19.0/bin/node
      Yarn: 1.22.5 - /usr/bin/yarn
      npm: 6.14.8 - ~/.nvm/versions/node/v12.19.0/bin/npm
    npmPackages:
      expo: ~40.0.0 => 40.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native-web: ~0.13.12 => 0.13.18 
    Expo Workflow: managed
```

**Please specify your device/emulator/simulator platform, model and version**

no specific device

**Error output**

```
wait  - compiling...
error - ./node_modules/@expo/next-adapter/document.js
SyntaxError: /home/Test/my-react-app/node_modules/@expo/next-adapter/document.js: Missing class properties transform.
  54 |
  55 | export class Document extends NextDocument {
> 56 |   static getInitialProps = getInitialProps;
     |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  57 |   render() {
  58 |     return (
  59 |       <Html>
```

**Reproducible demo or steps to reproduce from a blank project**

Create a new project

```
npx create-react-native-app -t with-nextjs --use-npm
```

Update NextJS to >=10.0.2

`npm i next@^10`

Start the dev server

`npx next dev`

While not a bug (@expo/next-adapter has a peer dependacy of ^9), it would be great if this library could exit beta supporting the latest major version of NextJS.

Versions 10.0.0 and 10.0.1 work without any issue, but versions >=10.0.2 have the error listed above.

This error only occurs when using the NextJS server (eg starting with `next dev`). When started using `expo start --web` it will compile with no errors.

**Comment by jonjamz** [comment]
While this has been working for me with 10.0.6, someone on my team was unable to run it until we locked it to 10.0.1 as you mentioned.

**Comment by marklawlor** [comment]
> While this has been working for me with 10.0.6, someone on my team was getting this same error and locking the version to 10.0.1 fixed it for them as well.

@jonjamz I've seen this behaviour as well and I think this has something to do with NextJS caching the built `_document.tsx` that was generated with 10.0.1. As the contents of the file hasn't changed will use continue to use this if you update to >=10.0.2. 
 I haven't tested this but presumably if you commit the built `_document.tsx` and committed the built file you could continue to >=10.0.2.

We have been using >=10.0.2 for some time and only noticed this behaviour after resetting out CI cache and testing some clean installs.

**Comment by louisgv** [comment]
> I think this has something to do with NextJS caching the built `_document.tsx` that was generated with 10.0.1. As the contents of the file hasn't changed will use continue to use this if you update to >=10.0.2.

This also caught me off-guard as well. I had 2 CI environments and because we cache next for incremental build, 1 CI succeed while the other one failed. The one failed doesn't have the cache.

**Comment by marklawlor** [comment]
> > I think this has something to do with NextJS caching the built `_document.tsx` that was generated with 10.0.1. As the contents of the file hasn't changed will use continue to use this if you update to >=10.0.2.
> 
> This also caught me off-guard as well. I had 2 CI environments and because we cache next for incremental build, 1 CI succeed while the other one failed. The one failed doesn't have the cache.
> 
> Found a dirty fix is to have the following in babel.config.js:
> 
> ```
> module.exports = {
>   presets: ["@expo/next-adapter/babel"],
>   plugins: [
>     "@emotion",
>     ["@babel/plugin-proposal-class-properties", { loose: true }]
>   ]
> }
> ```
> 
> If you really desperate for next@latest. For my case, I will be locking my next to 10.0.1

I'll test this later, but if that works you should be able to selectively only apply that fix to only the `_document.tsx` file

```
module.exports = {
  presets: ["@expo/next-adapter/babel"],
  overrides: [
    {
      test: "./pages/_document.tsx",
      plugins: [["@babel/plugin-proposal-class-properties", { loose: true }]]
    },
  ],
}

**Comment by tamagokun** [comment]
same thing happening to me on next 10.0.8, only it's happening with a random module in `react-native-animatable` that uses a static class prop. I can verify that `@babel/plugin-proposal-class-properties` is in node_modules.

Feels like something is wonky with the module resolution. Usually the fix is to nuke yarn.lock and reinstall, but recently that didn't cut it, and I had to add the plugin into our babel.config.js.

**Comment by marklawlor** [comment]
@tamagokun I just updated my previous comment, I've successfully updated to 10.0.8 by adding babel overrides for all my files that need the class transform.

If your having issues with `react-native-animatable` maybe if you update your babel config to something like this it will work

```
module.exports = {
  presets: ["@expo/next-adapter/babel"],
  overrides: [
    {
      test: "./node_modules/@expo/next-adapter/document.js",
      plugins: [["@babel/plugin-proposal-class-properties"]],
    },
    {
      test: "./node_modules/react-native-animatable/*",
      plugins: [["@babel/plugin-proposal-class-properties"]],
    },
  ],
}
```

**Comment by tamagokun** [comment]
yeah, adding the plugin manually or using overrides gets it to work right. Wish I could figure out WHY things are getting confused

**Comment by nandorojo** [comment]
I solved this by adding `@expo/next-adapter` to my transpile modules array in `next.config.js`.

```js
// next.config.js

const { withExpo } = require("@expo/next-adapter");

const withTM = require("next-transpile-modules")([
  "moti",
  "@motify/core",
  "@motify/components",
  "@expo/next-adapter",
]);

const withPlugins = require("next-compose-plugins");

module.exports = withPlugins([
  withTM,
  [
    withExpo,
    {
      projectRoot: __dirname,
    },
  ],
]);
```

Seems like a bit of a hack, since this won't work if you want to make your own document. It only works if you re-export the document from next adapter. This also solved it for me:

```js
// babel.config.js
module.exports = function (api) {
  api.cache(true);

  return {
    presets: ["@expo/next-adapter/babel"],
    plugins: [
      "react-native-reanimated/plugin",
      "@babel/plugin-proposal-class-properties",
    ],
  };
};
```

**Comment by vrinch** [comment]
this issue still persists
the solutions proffered here dint work for me tho
asides downgrading

**Comment by marklawlor** [comment]
Here is a repo using Next 10.2 w/ Webpack 5 https://github.com/marklawlor/expo-nextjs-webpack-5

If your having issues you can use it as an example. It also contains other fixes to allow you to upgrade

**Comment by louisgv** [comment]
> Here is a repo using Next 10.2 w/ Webpack 5 https://github.com/marklawlor/expo-nextjs-webpack-5
> 
> If your having issues you can use it as an example. It also contains other fixes to allow you to upgrade

Great work 👍

**Comment by marklawlor** [comment]
I'm closing this as NextJS just released v11. I've created a new issue #3579 to track all issues regarding full Next 11 support.

https://github.com/marklawlor/expo-nextjs-webpack-5 has also been updated for Next 11

**Comment by vrinch** [comment]
Still the same issue with next >=11

**Comment by marklawlor** [comment]
@vrinch The next-adapter 3.x removed its dependency on this transform. If you still require this transform for other 3rd party modules you will need to add it yourself.

Note that Expo never added this transform. It was part of the next babel preset and was removed in 10.x

**Comment by nandorojo** [comment]
Recently started getting this issue again in Next.js 12.2.5. I fixed it by using SWC instead of Babel, in case that's useful.

If you're using Reanimated, you can now use their Next.js SWC Plugin on Next.js 12.2 +. You can find it on the Showtime Repo.

**Comment by Raj210Kumari** [comment]
SyntaxError: C:\Users\Intel\Desktop\New folder\geekbuying\-acceptable-popcorn-8320\electrohippo\node_modules\@chakra-ui\modal\dist\index.esm.js: Missing class properties transform.
 29 | import { useEffect } from "react";
  30 | var ModalManager = class {
> 31 |   modals;
     |   ^^^^^^^
  32 |   constructor() {
  33 |     this.modals = [];
  34 |   }
    at transformFile.next (<anonymous>)
    at run.next (<anonymous>)
    at transform.next (<anonymous>)


Please help me resolve this chakra UI issue.

**Comment by nhydv** [comment]
> SyntaxError: C:\Users\Intel\Desktop\New folder\geekbuying-acceptable-popcorn-8320\electrohippo\node_modules@chakra-ui\modal\dist\index.esm.js: Missing class properties transform. 29 | import { useEffect } from "react"; 30 | var ModalManager = class {
> 
> > 31 |   modals;
> > |   ^^^^^^^
> > 32 |   constructor() {
> > 33 |     this.modals = [];
> > 34 |   }
> > at transformFile.next ()
> > at run.next ()
> > at transform.next ()
> 
> Please help me resolve this chakra UI issue.

same issue here. 
can any one help?

**Comment by nandorojo** [comment]
For some reason, this is happening to me again with `@expo/vector-icons`. I'm on Next.js 13.0.5. I tried with Babel + the class properties plugin, and with SWC. Happens with both. Other libraries that had this issue got fixed when added to Next Transpile Modules. For some reason, adding `@expo/vector-icons` to `next-transpile-modules` isn't fixing it.

I'm going to see if using Next's new built-in transpilation helps.
